runtime/src/accounts_db.rs

```diff
@@ -281,14 +281,62 @@ impl CumulativeOffsets1D {
         }
     }
 
+    pub fn from_raw_2d<T>(raw: &[Vec<Vec<T>>]) -> CumulativeOffsets {
+        let mut total_count: usize = 0;
+        let cumulative_offsets: Vec<CumulativeOffset> = raw
+            .iter()
+            .enumerate()
+            .map(|(i, v)| {
+                v.iter()
+                    .enumerate()
+                    .filter_map(|(j, v)| {
+                        let len = v.len();
+                        if len > 0 {
+                            let result = CumulativeOffset::new(vec![i, j], total_count);
+                            total_count += len;
```

Comment:
With the 2d version, this might be cleaner with the iterative version because it ends up with less code and I believe the inner `collect` allocates which is just then copied into the outer `collect` vec, so it should be slightly more efficient without that copy.

```
let cumulative_offsets = Vec::with_capacity(raw.len() * raw[0].len());
for (i, v) in raw.iter().enumerate() {
   for (j, v) in v.iter().enumerate() {
       let len = v.len();
       if len > 0 {
           cumulative_offsets.push(CumulativeOffset::new(vec![i, j], total_count));
           total_count += len;
       }
   }
}
cumulative_offsets 
```

Reply:
will address separately

Reply:
#15346